freqtrade/plugins/pairlist/MarketCapPairList.py

```diff
@@ -0,0 +1,178 @@
+        return {
+            "number_assets": {
+                "type": "number",
+                "default": 30,
+                "description": "Max market cap rank",
+                "help": "Only use assets with high market cap rank",
+            },
+            "mode": {
+                "type": "option",
+                "default": "top_rank",
+                "options": MODE_VALUES,
+                "description": "Mode of number",
+                "help": "How to interpret the number",
+            },
+            "refresh_period": {
+                "type": "number",
+                "default": 86400,
+                "description": "Refresh period",
+                "help": "Refresh period in seconds",
+            }
+        }
+
+    def gen_pairlist(self, tickers: Tickers) -> List[str]:
+        """
+        Generate the pairlist
+        :param tickers: Tickers (from exchange.get_tickers). May be cached.
+        :return: List of pairs
+        """
+        # Generate dynamic whitelist
+        # Must always run if this pairlist is the first in the list.
+        pairlist = self._marketcap_cache.get('pairlist_mc')
+        if pairlist:
+            # Item found - no refresh necessary
+            return pairlist.copy()
+        else:
+            # Use fresh pairlist
+            # Check if pair quote currency equals to the stake currency.
+            _pairlist = [k for k in self._exchange.get_markets(
+                quote_currencies=[self._stake_currency],
+                tradable_only=True, active_only=True).keys()]
+            # No point in testing for blacklisted pairs...
+            _pairlist = self.verify_blacklist(_pairlist, logger.info)
+
+            pairlist = self.filter_pairlist(_pairlist, tickers)
+            self._marketcap_cache['pairlist_mc'] = pairlist.copy()
+
+        return pairlist
+
+    def filter_pairlist(self, pairlist: List[str], tickers: Dict) -> List[str]:
+        """
+        Filters and sorts pairlist and returns the whitelist again.
+        Called on each bot iteration - please use internal caching if necessary
+        :param pairlist: pairlist to filter or sort
+        :param tickers: Tickers (from exchange.get_tickers). May be cached.
+        :return: new whitelist
+        """
+        marketcap_list = self._marketcap_cache.get('marketcap')
+
+        if marketcap_list is None:
+            data = self._coingekko.get_coins_markets(vs_currency='usd', order='market_cap_desc',
+                                                     per_page='250', page='1', sparkline='false',
+                                                     locale='en')
+            if data:
+                marketcap_list = [row['symbol'] for row in data]
+                self._marketcap_cache['marketcap'] = marketcap_list
+
+        if marketcap_list:
+            filtered_pairlist = []
+
+            if self._mode == 'top_rank':
+                top_marketcap = marketcap_list[:self._number_assets:]
```

Comment:
Are you sure that this is the correct approach (at least to be the default)?

running this with 30 - it's quite surprising that i only get about 20 pairs in the resulting pairlist ... ?

Reply:
Well, that's correct, as USDT is ranked 3rd for example. I believe your exchange don't have USDT/USDT pair. On which exchange do you try this? And do you use blacklist?

Reply:
i understand the reason :laughing: 
but i'm unsure if it's the correct approach (like, shouldn't there be 2 settings?).

assume i'd like to have 40 pairs ... but only from within the top 50 marketcap pairs (so if 12 are not available in my pairlist, my pairlist would only be 38 pairs long, otherwise, i'll always get 40 pairs).

right now, the 2 settings are mutually exclusive (changed via "mode") .. so i have to pick, that i _either_ get to have 40 pairs (total_assets mode) - _or_ pairs from within the top 50 marketcap pairs (top_rank mode).

it's _probably_ possible via some combination of 2 sequential pairlists (didn't test, hence probably) - but with coingecko, that almost certainly isn't good idea, as it'd mean we have 3 coingecko objects around - doing 2x the same request in series (each pairlist is it's own instance) (the 3rd is fiatconvert).

In my initial understanding (before looking at the code in detail) was that `mode` would switch the sorting of the coingecko pairs - essentially one of `market_cap_asc, market_cap_desc, volume_asc, volume_desc, id_asc, id_desc` (taken from the API endpoint docs). if that makes sense is a diff. discussion.

an alternative approach to get the above scenario (and be more flexible overall) could be the following:
* `number_assets` is optional - defining the max length of the pairlist (as it's handled in most other pairlists)
* new setting "marketcap_top" (or whatever better name) defines the marketcap "top" pairs (essentially how many coingecko pairs we'll consider).

that 2nd setting is capped at 250 - and lower values will obviously mean there's fewer values to pick from.

Reply:
> assume i'd like to have 40 pairs ... but only from within the top 50 marketcap pairs (so if 12 are not available in my pairlist, my pairlist would only be 38 pairs long, otherwise, i'll always get 40 pairs).

You can do this using stacked filter. First one on `total_assets` mode using 40 `number_assets`. Then second one on `top_rank` mode on 50 `number_assets`. I should add this example in the docs.

> it's probably possible via some combination of 2 sequential pairlists (didn't test, hence probably) - but with coingecko, that almost certainly isn't good idea, as it'd mean we have 3 coingecko objects around - doing 2x the same request in series (each pairlist is it's own instance) (the 3rd is fiatconvert).

Well, the cache is 1d cache, and in theory, assuming the cache expired, only the first pairlist call the API, the second pairlist will use the cache. Don't think it would work differently compared to stacked volumepairlist. Gonna add a log_once when the bot fetch data, so we would know if there is a rare case of double API call

Reply:
> Well, the cache is 1d cache, and in theory, assuming the cache expired, only the first pairlist call the API, the second pairlist will use the cache

that's incorrect i'm afraid.
The cache is per instance - not a global cache - so if you specify 20 MarketCapPairlists in sequence, then you'll do 20 calls to coingecko (20 identical calls, but still), as each entry into pairlists is a seperate, independent instance.

Reply:
ah okay. Let's see if I can change the implementation

Reply:
i'm ... not sure if turning this into a global (so cache applies for everything) is a good idea.
there's quite a few modes (webserver -> test-pairlist for example) where this can greatly mess up behavior...